File: src/main/java/liqp/tags/For.java
package liqp.tags;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import liqp.LValue;
import liqp.exceptions.ExceededMaxIterationsException;
import liqp.nodes.BlockNode;
import liqp.nodes.LNode;
import liqp.nodes.RenderContext;

class For extends Tag {

    private static final String OFFSET = "offset";
    private static final String LIMIT = "limit";
    private static final String CONTINUE = "continue";

    /*
     * forloop.length      # => length of the entire for loop
     * forloop.index       # => index of the current iteration
     * forloop.index0      # => index of the current iteration (zero based)
     * forloop.rindex      # => how many items are still left?
     * forloop.rindex0     # => how many items are still left? (zero based)
     * forloop.first       # => is this the first iteration?
     * forloop.last        # => is this the last iteration?
     */

  /*
   * For loop
   */
  @Override
  public Object render(RenderContext context, LNode... nodes) {

    // The first node in the array denotes whether this is a for-tag
    // over an array, `for item in array ...`, or a for-tag over a
    // range, `for i in (4..item.length)`.
    boolean array = super.asBoolean(nodes[0].render(context));

    String id = super.asString(nodes[1].render(context));

    // Each for tag has its own context that keeps track of its own variables (scope)
    context.addFrame();

    Object rendered = array ? renderArray(id, context, nodes) : renderRange(id, context, nodes);

    context.popFrame();
    return rendered;
  }

  private Object renderArray(String id, RenderContext context, LNode... tokens) {

    StringBuilder builder = new StringBuilder();

    // attributes start from index 5
    Map<String, Integer> attributes = getAttributes(5, context, tokens);

    int offset = attributes.get(OFFSET);
    int limit = attributes.get(LIMIT);

    Object[] array = super.asArray(tokens[2].render(context));

    LNode block = tokens[3];
    LNode blockIfEmptyOrNull = tokens[4];

    if (array == null || array.length == 0) {
      return blockIfEmptyOrNull == null ? null : blockIfEmptyOrNull.render(context);
    }

    int length = Math.min(limit, array.length);

    final String loopName = id + "-" + tokens[2].toString();
    context.startLoop(length, loopName);

    int continueIndex = offset;

    for (int i = offset, n = 0; n < limit && i < array.length; i++, n++) {

      context.incrementIterations();

      continueIndex = i;
      context.set(id, array[i]);

      List<LNode> children = ((BlockNode) block).getChildren();
      boolean isBreak = false;

      for (LNode node : children) {

        Object value = node.render(context);

        if (value == LValue.CONTINUE) {
          // break from this inner loop: equals continue outer loop!
          break;
        }

        if (value == LValue.BREAK) {
          // break from inner loop
          isBreak = true;
          break;
        }

        if (value != null && value.getClass().isArray()) {

          Object[] arr = (Object[]) value;

          for (Object obj : arr) {
            builder.append(String.valueOf(obj));
          }
        } else {
          builder.append(super.asString(value));
        }
      }

      if (isBreak) {
        // break from outer loop
        break;
      }
    }

    context.setRoot(CONTINUE, continueIndex + 1);

    return builder.toString();
  }

  private Object renderRange(String id, RenderContext context, LNode... tokens) {

    StringBuilder builder = new StringBuilder();

    // attributes start from index 5
    Map<String, Integer> attributes = getAttributes(5, context, tokens);

    int offset = attributes.get(OFFSET);
    int limit = attributes.get(LIMIT);

    LNode block = tokens[4];

    try {
      int from = super.asNumber(tokens[2].render(context)).intValue();
      int to = super.asNumber(tokens[3].render(context)).intValue();

      int length = (to - from);

      context.startLoop(length, null);
      int continueIndex = from + offset;

      for (int i = from + offset, n = 0; i <= to && n < limit; i++, n++) {

        context.incrementIterations();
        continueIndex = i;
        context.set(id, i);

        List<LNode> children = ((BlockNode) block).getChildren();
        boolean isBreak = false;

        for (LNode node : children) {

          Object value = node.render(context);

          if (value == null) {
            continue;
          }

          if (value == LValue.CONTINUE) {
            // break from this inner loop: equals continue outer loop!
            break;
          }

          if (value == LValue.BREAK) {
            // break from inner loop
            isBreak = true;
            break;
          }

          if (super.isArray(value)) {

            Object[] arr = super.asArray(value);

            for (Object obj : arr) {
              builder.append(String.valueOf(obj));
            }
          } else {
            builder.append(super.asString(value));
          }
        }

        if (isBreak) {
          // break from outer loop
          break;
        }
      }

      context.set(CONTINUE, continueIndex + 1);
    } catch (ExceededMaxIterationsException e) {
      throw e;
    } catch (Exception e) {
      /* just ignore incorrect expressions */
    }

    return builder.toString();
  }

  private Map<String, Integer> getAttributes(int fromIndex, RenderContext context, LNode... tokens) {

    Map<String, Integer> attributes = new HashMap<String, Integer>();

    attributes.put(OFFSET, 0);
    attributes.put(LIMIT, Integer.MAX_VALUE);

    for (int i = fromIndex; i < tokens.length; i++) {

      Object[] attribute = super.asArray(tokens[i].render(context));

      try {
        attributes.put(super.asString(attribute[0]), super.asNumber(attribute[1]).intValue());
      } catch (Exception e) {
        /* just ignore incorrect attributes */
      }
    }

    return attributes;
  }
}

File: liqp-core/src/main/kotlin/liqp/tags/For.kt
package liqp.tags

import liqp.ControlResult.BREAK
import liqp.ControlResult.CONTINUE
import liqp.context.LContext
import liqp.node.LNode
import liqp.params.ResolvableNamedParams
import liqp.safeSlice
import liqp.tag.LTag

class For : LTag() {

  /**
   * forloop.length      # => length of the entire for loop
   * forloop.index       # => index of the current iteration
   * forloop.index0      # => index of the current iteration (zero based)
   * forloop.rindex      # => how many items are still left?
   * forloop.rindex0     # => how many items are still left? (zero based)
   * forloop.first       # => is this the first iteration?
   * forloop.last        # => is this the last iteration?
   */
  override fun render(context: LContext, vararg nodes: LNode): Any? {

    // The first node in the array denotes whether this is a for-tag
    // over an array, `for item in array ...`, or a for-tag over a
    // range, `for i in (4..item.length)`.
    val isArray:Boolean = nodes[0].execute(context)
    val id:String = nodes[1].execute(context)


    // attributes start from index 5
    val attributes = ResolvableNamedParams(context,
        paramNodes = nodes.slice(5..nodes.lastIndex)
    )

    val offset:Int = attributes[OFFSET, 0]
    val limit:Int = attributes[LIMIT, Integer.MAX_VALUE]

    // Each for tag has its own context that keeps track of its own variables (scope)
    return context.withFrame {
      when (isArray) {
        true -> renderArray(id, context, offset, limit, *nodes)
        false -> renderRange(id, context, offset, limit, *nodes)
      }
    }
  }

  private fun renderArray(id: String, context: LContext,
                          offset:Int, limit:Int,
                          vararg tokens: LNode): Any? {

    val results = mutableListOf<Any?>()

    val iterated:List<Any> = tokens[2].executeOrNull(context) ?: emptyList()

    val block = tokens[3]
    val blockIfEmptyOrNull = tokens[4]

    if (iterated.isEmpty()) {
      return blockIfEmptyOrNull.render(context)
    }

    val length = Math.min(limit, iterated.size)

    val loopName = id + "-" + tokens[2].toString()
    context.startLoop(length, loopName)

    var continueIndex = offset

    val adjustedLimit = when (limit) {
      Integer.MAX_VALUE -> iterated.size
      else -> offset + limit
    }

    var i = offset
    val sliced = iterated.safeSlice(offset, adjustedLimit)
    outer@for (item in sliced) {
      context.incrementIterations()

      continueIndex = i
      context[id] = item

      val children = block.children
      inner@for (node in children) {
        val value = node.render(context)

        if (value == CONTINUE) {
          // break from this inner loop: equals continue outer loop!
          break@inner
        }

        if (value == BREAK) {
          break@outer
        }

        results.addAll(context.asIterable(value))
      }
      i++
    }

    context.setRoot(CONTINUE.name.toLowerCase(), continueIndex + 1)
    return results
  }

  private fun renderRange(id: String, context: LContext, offset:Int, limit:Int, vararg tokens: LNode): Any {

    val results = mutableListOf<Any?>()
    val block = tokens[4]

    val from:Int = tokens[2].executeOrNull<Int>(context) ?: return listOf<Any>()
    val to:Int = tokens[3].execute(context, 0)

    val length = to - from

    context.startLoop(length, null)
    var continueIndex = from + offset

    var i = from + offset
    var n = 0
    outer@while (i <= to && n < limit) {

      context.incrementIterations()
      continueIndex = i
      context[id] = i

      val children = block.children
      inner@for (node in children) {

        val value = node.render(context) ?: continue@inner

        if (value === CONTINUE) {
          // break from this inner loop: equals continue outer loop!
          break@inner
        }

        if (value === BREAK) {
          // break from outer loop
          break@outer
        }

        if (context.isIterable(value)) {
          results.addAll(context.asIterable(value))
        } else {
          results.add(value)
        }
      }
      i++
      n++
    }

    context[CONTINUE.name] = continueIndex + 1


    return results
  }
}
